### PR TITLE
NBT: Add support for nonarray lists, fix bugs with structs, make compliant with bigtest.nbt

### PR DESCRIPTION
**Pull request by masp**
Fixes:
- Bugs with structs and interfaces I added (woopsie!)
- Adds compliance test for bigtest.nbt that wiki.vg says is standard

Adds:
- nbt_type tag to allow `[]uint64` to be encoded as list of int64 rather than int64 array. This was to support bigtest.nbt since it has a list of longs rather than an array of longs.
- MarshalCompound(...) function to make it so you can create a root NBT compound with a name

Example:

```go
type Test struct {
    NotArray  []uint64 `nbt:"This is a list of TAG_Longs not an array" nbt_type:"notarray"`
}
```

### CODE REVIEW

File: nbt/marshal.go [hunk 1]
```diff
@@ -21,166 +35,86 @@ func NewEncoder(w io.Writer) *Encoder {
...
+	return e.marshal(val, getTagType(val.Type()), "")
 }
 
-func (e *Encoder) marshal(val reflect.Value, tagName string) error {
-	switch vk := val.Kind(); vk {
-	default:
-		return errors.New("unknown type " + vk.String())
+func (e *Encoder) marshal(val reflect.Value, tagType byte, tagName string) (err error) {
+	err = e.writeHeader(val, tagType, tagName)
+	err = e.writeValue(val, tagType)
```
Comment (Tnze) [inline nbt/marshal.go:43]: You covered the last err here without checking.

**Comment by Tnze** [comment]
I prefer new unit test can share the bigtest file with `nbt_test.go`, so you don't need to put that `.nbt` file in the package alone. Please have a try~

**Comment by masp** [comment]
@Tnze Sorry about that! I missed that file completely.

**Comment by masp** [comment]
@Tnze The issues you mentioned were fixed if you want to look again.

**Comment by Tnze** [comment]
Thanks for this work~
One more thing, could you please add some document of new features you contributed? Like the usage of `nbt_type`.

**Pull request merged**